sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+
+Note that the post-link tool will generate a series of `OpSpecConstantOp` ops
+when the kernel or exported device function has multiple requirements.
+However, each property list entry contains only the `SpecId` of the
+`OpSpecConstantTrue` op that is associated with a single requirement.
+
+#### Modifications to the DPC++ runtime
+
+Modifications are also required to the DPC++ runtime in order to set the values
+of the specialization constants that correspond to each kernel requirement.  In
+addition, the runtime needs to check if the target device supports each of the
+kernel's requirements, and it must raise an `errc::kernel_not_supported`
+exception if it does not.
+
+When a kernel is submitted to a device, the runtime finds the device image that
+contains the kernel and also finds the kernel's entry in the
+"SYCL/requirements" property set.  This entry tells the set of requirements for
+the kernel.  If the target device does not support all of these requirements,
+then the runtime raises `errc::kernel_not_supported`.  This check can be done
+before the device image is JIT compiled, so the exception can be thrown
+synchronously.
+
+If the kernel imports device function symbols from a shared library as defined
+in [Device Code Dynamic Linking][7], the runtime first identifies all the
+device images that define these exported device functions.  Before attempting
+to link them together, the runtime finds the entries for the exported device
+functions in their "SYCL/requirements" property sets and checks that the device
+supports all these requirements.  If it does not, the runtime throws
+`errc::kernel_not_supported`.
+
+Whenever the runtime submits a SPIR-V image to the backend for online
+compilation, it must do the following additional steps:
+
+* Compute the union of all requirements from all entries in the image's
+  "SYCL/requirements" property set.
+
+* Query the target device to see whether it supports each of these
+  requirements, yielding either `True` or `False` for each one.
+
+* Set the value of each associated specialization constant when JIT compiling
+  the device image for this target device.
+
+Note that the runtime's cache of compiled device images does not need any
+special modification because the cache already needs to know the values of all
+the specialization constants that were used to compile the device image.  We
+just need to make sure the cache is also aware of the specialization constants
+which correspond to the requirements from the "SYCL/requirements" property set.
+
+#### Modifications to the GEN compiler
+
+The GEN compiler, of course, needs to be modified to implement the new
+`ConditionalINTEL` SPIR-V decoration.  It must discard any function with this
+decoration (unless the corresponding specialization constant is `True`), and it
+must not raise any sort of error due to compilation of these discarded
+functions.
```

Comment:
Note: SPIR-V specialization step is performed by the SPIR-V translator and therefore shared between all the backends using it:

https://github.com/intel/llvm/blob/611746cb9b4d099b55a42a5043aa3114ceacec30/llvm-spirv/tools/llvm-spirv/llvm-spirv.cpp#L157-L165

This section could probably be omitted in favor of saying that functions are discarded during [Specialization](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_specializationsection_a_specialization) step in the corresponding SPIR-V extension step

Reply:
> I presume the SPIRV->LLVM translator is incorporated into the OpenCL and Level Zero drivers, right?

That is an implementation detail. I think the best way here would be to say that we _require_ backends to support `SPV_INTEL_spec_conditional`. This automatically will imply discarding "disabled" functions and integration of the corresponding SPIR-V translator patch (in some or another way).

BTW, do we really *require* this extension now for a pretty simple SYCL applications which can be otherwise represented using core SPIR-V functionality?

Reply:
> BTW, do we really require this extension now for even the simplest SYCL applications which can be otherwise represented using core SPIR-V functionality?

I think we should not say that the driver is required to implement all of `SPV_INTEL_spec_conditional` since many of those features are intended to support the [device-if](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc) feature.  Instead, we could say that the driver is required to support only the `ConditionalINTEL` decoration from `SPV_INTEL_spec_conditional`.

Reply:
> Instead, we could say that the driver is required to support only the ConditionalINTEL decoration from SPV_INTEL_spec_conditional.

Still, this reduces portability: in order to add your backend you will have to support some non-standard `INTEL` extension, not even `KHR`-one. BTW, what about non-SPIR-V backends? How conditional compilation intended to work there?

Reply:
> Still, this reduces portability: in order to add your backend you will have to support some non-standard INTEL extension, not even KHR-one. BTW, what about non-SPIR-V backends? How conditional compilation intended to work there?

There is a big design decision to make here:

1. We can bundle all kernels together into device images as we normally do, but decorate the SPIR-V in a way that allows some functions to be discarded at JIT compilation time.  This is the strategy I describe now in this design document.  Or,

2. We can change the bundling strategy in the post-link tool so that kernels are only bundled together into the same device image if they use exactly the same set of device aspects.  This is an alternate design that I allude to in the paragraph that starts "There are two ways to solve this problem ...".

Initially, I thought (2) was the best approach because it didn't require any SPIR-V extensions.  But, then I realized that we would need the SPIR-V extension anyways in order to implement the [device-if](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc) feature.

You are correct that option (2) would be easier to port to different backends because it doesn't require any SPIR-V extension.  It would also be more straightforward to port to backends that don't use SPIR-V.  I think the main disadvantage to (2) is that it's extra work, especially if we plan to implement [device-if](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc).  Assuming that we do implement device-if using the SPIR-V extension, it's relatively easy to use that same extension to implement option (1).

Answering your question about non-SPIR-V backends, we would have to take this on a case-by-case basis.  If the backend supported some sort of way to conditionally compile code at JIT time, we could use that feature.  If it did not, we could always implement option (2) just for that backend (and for any similar backends).

I know from previous discussions that @kbobrovs prefers option (1), and I think @kbsmith-intel has expressed a preference for that also.  We did not specifically discuss the flexibility to support non-SPIR-V backends, though.  I think we need to decide if the extra effort required for option (2) is worth the extra flexibility it provides.

Reply:
We had an offline meeting and decided to go with design option (2).  This design document has already been updated to reflect that choice.